collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -1255,6 +1063,216 @@ public void onClick(View v) {
 		}
 	}
 
+	@NonNull
+	private View createFormEndView(FormController formController) {
+		View endView = View.inflate(this, R.layout.form_entry_end, null);
+		((TextView) endView.findViewById(R.id.description))
+                .setText(getString(R.string.save_enter_data_description,
+                        formController.getFormTitle()));
+
+		// checkbox for if finished or ready to send
+		final CheckBox instanceComplete = ((CheckBox) endView
+                .findViewById(R.id.mark_finished));
+		instanceComplete.setChecked(isInstanceComplete(true));
+
+		if (!mAdminPreferences.getBoolean(
+                AdminPreferencesActivity.KEY_MARK_AS_FINALIZED, true)) {
+            instanceComplete.setVisibility(View.GONE);
+        }
+
+		// edittext to change the displayed name of the instance
+		final EditText saveAs = (EditText) endView
+                .findViewById(R.id.save_name);
+
+		// disallow carriage returns in the name
+		InputFilter returnFilter = new InputFilter() {
+            public CharSequence filter(CharSequence source, int start,
+									   int end, Spanned dest, int dstart, int dend) {
+                for (int i = start; i < end; i++) {
```

Comment:
Some strange things happened to indentation and line breaks throughout these new methods!